precert/corr.go

```diff
@@ -0,0 +1,202 @@
+		return fmt.Errorf("no SCTList extension found in final cert")
+	}
+	if finalCertParser.skipped > 1 {
+		return fmt.Errorf("multiple SCTList extensions found in final cert")
+	}
+	return nil
+}
+
+var poisonOID = []int{1, 3, 6, 1, 4, 1, 11129, 2, 4, 3}
+var sctListOID = []int{1, 3, 6, 1, 4, 1, 11129, 2, 4, 2}
+
+// extensionParser takes a sequence of bytes representing the inner bytes of the
+// `extensions` field. Repeated calls to Next() will return all the extensions
+// except those that match the skippableOID. The skipped extensions will be
+// counted in `skipped`.
+type extensionParser struct {
+	skippableOID encoding_asn1.ObjectIdentifier
+	bytes        cryptobyte.String
+	skipped      int
+}
+
+// Next returns the next extension in the sequence, skipping (and counting)
+// any extension that matches the skippableOID.
+// Returns nil, nil when there are no more extensions.
+func (e *extensionParser) Next() (cryptobyte.String, error) {
+	if e.bytes.Empty() {
+		return nil, nil
+	}
+
+	var next cryptobyte.String
+	if !e.bytes.ReadASN1(&next, asn1.SEQUENCE) {
+		return nil, fmt.Errorf("failed to parse extension")
+	}
+
+	var oid encoding_asn1.ObjectIdentifier
+	nextCopy := next
+	if !nextCopy.ReadASN1ObjectIdentifier(&oid) {
+		return nil, fmt.Errorf("failed to parse extension OID")
+	}
+
+	if oid.Equal(e.skippableOID) {
+		e.skipped++
+		return e.Next()
+	}
+
+	return next, nil
+}
+
+// unwrapExtensions takes a given a sequence of bytes representing the `extensions` field
+// of a TBSCertificate and parses away the outermost two layers, returning the inner bytes
+// of a SEQUENCE, which can then be parsed as a list of extensions.
+func unwrapExtensions(field cryptobyte.String) (cryptobyte.String, error) {
+	var extensions cryptobyte.String
+	if !field.ReadASN1(&extensions, asn1.Tag(3).Constructed().ContextSpecific()) {
+		return nil, errors.New("error reading extensions")
+	}
+
+	var extensionsInner cryptobyte.String
+	if !extensions.ReadASN1(&extensionsInner, asn1.SEQUENCE) {
+		return nil, errors.New("error reading extensions inner")
+	}
+
+	return extensionsInner, nil
+}
+
+// readIdenticalElement parses a single ASN1 element and returns an error if
+// their tags are different or their contents are different.
+func readIdenticalElement(a, b *cryptobyte.String) error {
+	var aInner, bInner cryptobyte.String
+	var aTag, bTag asn1.Tag
+	if !a.ReadAnyASN1Element(&aInner, &aTag) {
+		return fmt.Errorf("failed to read element from first input")
+	}
+	if !b.ReadAnyASN1Element(&bInner, &bTag) {
+		return fmt.Errorf("failed to read element from first input")
+	}
+	if aTag != bTag {
+		return fmt.Errorf("tags differ: %d != %d", aTag, bTag)
+	}
+	if !bytes.Equal([]byte(aInner), []byte(bInner)) {
+		return fmt.Errorf("elements differ: %x != %x", aInner, bInner)
+	}
+	return nil
+}
+
+// tbsDERFromCertDER takes a Certificate object encoded as DER, and parses
+// away the outermost two SEQUENCEs to get the TBSCertificate.
```

Comment:
IMO it would be useful to show the ASN.1 definition that makes this correct, similar to how you show the ASN.1 of a tbsCertificate up on line 33.

Same for `unwrapExtensions`.